test/jdk/javax/crypto/KeyGenerator/CompareKeys.java

```diff
@@ -0,0 +1,153 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import javax.crypto.SecretKey;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.security.Key;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.util.Arrays;
+import javax.crypto.KeyGenerator;
+
+/*
+ * @test
+ * @bug 8185127
+ * @summary Test key comparison between the Keys generated through KeyGenerator
+ * @run main CompareKeys
+ */
+public class CompareKeys {
+
+    public static void main(String[] args) throws Exception {
+
+        for (KeyAlgo ka : KeyAlgo.values()) {
+            SecretKey k = ka.genSecretKey();
+            checkKeyEquality(k, copy(k));
+        }
+
+        for (KeyPairAlgo kpa : KeyPairAlgo.values()) {
+            KeyPair kp = kpa.genKeyPair();
+            checkKeyPairEquality(kp, copy(kp));
+        }
+        System.out.println("Done!");
+    }
+
+    private static void checkKeyPairEquality(KeyPair origkp, KeyPair copykp)
+            throws Exception {
+
+        checkKeyEquality(origkp.getPrivate(), copykp.getPrivate());
+        checkKeyEquality(origkp.getPublic(), copykp.getPublic());
+    }
+
+    /**
+     * Compare original Key with another copy.
+     */
+    private static void checkKeyEquality(Key origKey, Key copyKey) {
+
+        if (!origKey.equals(copyKey)
+                && !Arrays.equals(origKey.getEncoded(), copyKey.getEncoded())
+                && origKey.hashCode() != copyKey.hashCode()) {
+            throw new RuntimeException("Key inequality found");
+        }
+        System.out.printf("Equality check Passed for key Type:%s & Algo: %s%n",
+                origKey.getClass(), origKey.getAlgorithm());
+    }
+
+    /**
+     * Get a copy of the original object type.
+     */
+    private static <T extends Object> T copy(T orig)
+            throws IOException, ClassNotFoundException {
+
+        byte[] serialize;
+        try ( ByteArrayOutputStream bos = new ByteArrayOutputStream();
+                ObjectOutputStream oos = new ObjectOutputStream(bos)) {
+            oos.writeObject(orig);
+            serialize = bos.toByteArray();
+        }
+
+        T copy;
+        try ( ByteArrayInputStream bis = new ByteArrayInputStream(serialize);
+                ObjectInputStream ois = new ObjectInputStream(bis)) {
+            copy = (T) ois.readObject();
+        }
+        return copy;
+    }
+
+    private enum KeyAlgo {
+
+        AES("AES"),
+        ARCFOUR("ARCFOUR"),
+        Blowfish("Blowfish"),
+        ChaCha20("ChaCha20"),
+        DES("DES"),
+        DESede("DESede"),
+        HmacMD5("HmacMD5"),
+        HmacSHA1("HmacSHA1"),
+        HmacSHA224("HmacSHA224"),
+        HmacSHA256("HmacSHA256"),
+        HmacSHA384("HmacSHA384"),
+        HmacSHA512("HmacSHA512"),
+        RC2("RC2");
```

Comment:
Just curious, how are these decided? Should this be an exhaustive list or just enough sampling for code coverage? If this is meant to be a general test, have you tried to not hardcoding the algorithm names to be enum? Otherwise, new algorithms will not be tested if not updating this test.

Reply:
Done. Key(pair)Generator list will be collected dynamically.

Reply:
Sounds good.